src/emqx_ws_connection.erl

```diff
@@ -183,16 +183,44 @@ init(Req, Opts) ->
                max_frame_size => MaxFrameSize,
                idle_timeout   => IdleTimeout
               },
-    case cowboy_req:parse_header(<<"sec-websocket-protocol">>, Req) of
+
+    case check_origin_header(Req, Opts) of
+        {error, Message} ->
+            {ok, cowboy_req:reply(400, Req), WsOpts};
+        ok ->
+            case cowboy_req:parse_header(<<"sec-websocket-protocol">>, Req) of
+                undefined ->
+                    %% TODO: why not reply 500???
+                    {cowboy_websocket, Req, [Req, Opts], WsOpts};
+                [<<"mqtt", Vsn/binary>>] ->
+                    Resp = cowboy_req:set_resp_header(
+                        <<"sec-websocket-protocol">>, <<"mqtt", Vsn/binary>>, Req),
+                    {cowboy_websocket, Resp, [Req, Opts], WsOpts};
+                _ ->
+                    {ok, cowboy_req:reply(400, Req), WsOpts}
+            end
+    end.
+
+parse_header_fun_origin(Req, Opts) ->
+    case cowboy_req:header(<<"origin">>, Req) of
         undefined ->
-            %% TODO: why not reply 500???
-            {cowboy_websocket, Req, [Req, Opts], WsOpts};
-        [<<"mqtt", Vsn/binary>>] ->
-            Resp = cowboy_req:set_resp_header(
-                     <<"sec-websocket-protocol">>, <<"mqtt", Vsn/binary>>, Req),
-            {cowboy_websocket, Resp, [Req, Opts], WsOpts};
-        _ ->
-            {ok, cowboy_req:reply(400, Req), WsOpts}
+                case proplists:get_value(allow_origin_absence, Opts, true) of
+                    true -> ok;
+                    false -> {error, origin_header_cannot_be_absent}
+                end;
+        Value ->
+            CheckOrigins = proplists:get_value(check_origins, Opts, ""),
+            Origins = [ list_to_binary(string:trim(O)) || O <- string:tokens(CheckOrigins, ",")],
```

Comment:
this is unnecessary string parsing repeated for each and every request.
maybe transform it in the config schema.